### PR TITLE
Add support for terry tasks.

### PR DESCRIPTION
**Pull request by veluca93**
Add an entry to the md files with name = whatever and terry = true to add a terry-style task.

### CODE REVIEW

File: components/category/Category.tsx [hunk 1]
```diff
@@ -10,12 +11,12 @@ import ReactMarkdown from "react-markdown";
 import React from "react";
 import { Progress } from "./Progress";
 
-function TaskList({ tasks }: { tasks: TaskScores }) {
-  const Task = ({ task, score }: { task: string; score: number }) => {
+function TaskList({ tasks , task_urls }: { tasks: TaskScores, task_urls: TaskURLs }) {
```
Comment (edomora97) [inline components/category/Category.tsx]:
```suggestion
function TaskList({ tasks, taskUrls }: { tasks: TaskScores, taskUrls: TaskURLs }) {
```

Use camelCase.

File: components/category/Category.tsx [hunk 2]
```diff
@@ -10,12 +11,12 @@ import ReactMarkdown from "react-markdown";
...
+  const Task = ({ task, score, url }: { task: string; score: number, url: string }) => {
     return (
       <li>
         <a
-          href={`https://training.olinfo.it/#/task/${task}/statement`}
+          href={`${url}`}
```
Comment (edomora97) [inline components/category/Category.tsx]:
```suggestion
          href={url}
```

File: lib/badges.ts [hunk 3]
```diff
@@ -14,12 +14,14 @@ export const GOLD_COLOR = "#ffdf00";
...
 
 export type TaskScores = { [taskId: string]: number };
+export type TaskURLs = { [taskId: string]: string };
 
 export type CategoryBadge = {
   node: Node;
   score: number;
   badge: Badge;
   tasks: TaskScores;
+  task_urls: TaskURLs;
```
Comment (edomora97) [inline lib/badges.ts]:
```suggestion
  taskUrls: TaskURLs;
```

File: lib/badges.ts [hunk 4]
```diff
@@ -73,16 +75,22 @@ export function computeCategoryBadges(
...
     let score = 0;
     const categoryTasks: TaskScores = {};
+    const categoryTaskURLs: TaskURLs = {};
     for (const task of node.tasks) {
       const taskScore = taskScores[task.name] ?? 0;
       categoryTasks[task.name] = taskScore;
+      categoryTaskURLs[task.name] = 'https://training.olinfo.it/#/task/' + task.name + '/statement';
+      if (task.terry) {
+        categoryTaskURLs[task.name] = 'https://territoriali.olinfo.it/task/' + task.name;
+      }
```
Comment (edomora97) [inline lib/badges.ts]:
```suggestion
      if (task.terry) {
        categoryTaskURLs[task.name] = 'https://territoriali.olinfo.it/task/' + task.name;
      } else {
        categoryTaskURLs[task.name] = 'https://training.olinfo.it/#/task/' + task.name + '/statement';
      }
```

File: lib/training-api.ts [hunk 5]
```diff
@@ -25,8 +25,17 @@ export async function getUserInfo(username: string): Promise<UserInfo | null> {
       credentials: "include",
     });
     if (req.status !== 200) return null;
-    const data = await req.json();
+    var data = await req.json();
```
Comment (edomora97) [inline lib/training-api.ts]:
```suggestion
    const data = await req.json();
```

File: lib/training-api.ts [hunk 6]
```diff
@@ -25,8 +25,17 @@ export async function getUserInfo(username: string): Promise<UserInfo | null> {
...
     if (data.success !== 1) return null;
+    const req_terr = await fetch("https://territoriali.olinfo.it/api/user/" + username + "/scores", {
```
Comment (edomora97) [inline lib/training-api.ts]:
```suggestion
    const reqTerr = await fetch(`https://territoriali.olinfo.it/api/user/${username}/scores`, {
```

File: lib/training-api.ts [hunk 7]
```diff
@@ -25,8 +25,17 @@ export async function getUserInfo(username: string): Promise<UserInfo | null> {
...
+      method: "GET",
+    });
+    if (req_terr.status != 200) return null;
```
Comment (edomora97) [inline lib/training-api.ts]: If the user doesn't exists on terry we don't want to fail the request also for training.

File: lib/training-api.ts [hunk 8]
```diff
@@ -25,8 +25,17 @@ export async function getUserInfo(username: string): Promise<UserInfo | null> {
       credentials: "include",
     });
     if (req.status !== 200) return null;
-    const data = await req.json();
+    var data = await req.json();
     if (data.success !== 1) return null;
```
Comment (edomora97) [inline lib/training-api.ts:30]: Note: an account on terry implies an account on training (with few negligible exceptions); short-circuiting the failure here is fine.

File: lib/training-api.ts [hunk 9]
```diff
@@ -25,8 +25,17 @@ export async function getUserInfo(username: string): Promise<UserInfo | null> {
...
+    const req_terr = await fetch("https://territoriali.olinfo.it/api/user/" + username + "/scores", {
+      method: "GET",
+    });
+    if (req_terr.status != 200) return null;
+    const terr_data = await req_terr.json();
+    for (var terr_task of terr_data) {
```
Comment (edomora97) [inline lib/training-api.ts]:
```suggestion
    for (const terrTask of terrData) {
```

File: lib/training-api.ts [hunk 10]
```diff
@@ -25,8 +25,17 @@ export async function getUserInfo(username: string): Promise<UserInfo | null> {
...
+      terr_task['score'] = terr_task['score'] * 100 / terr_task['max_score'];
```
Comment (edomora97) [inline lib/training-api.ts]:
```suggestion
      terr_task['score'] = terr_task['score'] * 50 / terr_task['max_score'];
```

We decided to make terry tasks worth 50 points.

**Review by edomora97 — CHANGES_REQUESTED** [review]

File: lib/training-api.ts (hunk 10 above)
Comment (veluca93) [reply lib/training-api.ts]: Other parts of the UI assume you can get up to 100 points in each task.
I think we can fix this later, and have them worth 100 points for now - it is sufficient for getting started on training for the regional competition.

**Comment by edomora97** [comment]
I'm fixing stuff. Wait before making other changes.

**Review by edomora97 — APPROVED** [review]

**Pull request merged**